cli/cmd/logs.go

```diff
@@ -0,0 +1,195 @@
+	if err != nil {
+		return nil, err
+	}
+	config.Template = tmpl
+
+	if c.labelSelector == "" {
+		config.LabelSelector = labels.Everything()
+	} else {
+		selector, err := labels.Parse(c.labelSelector)
+		if err != nil {
+			return nil, err
+		}
+		config.LabelSelector = selector
+	}
+
+	if c.tail != -1 {
+		config.TailLines = &c.tail
+	}
+
+	config.Since = c.sinceSeconds
+	config.Timestamps = c.timestamps
+	config.Namespace = c.namespace
+
+	return config, nil
+}
+
+func newLogOptions(flags *logFlags, kubeconfigPath, kubeContext string) (*logCmdOpts, error) {
+	kubeAPI, err := k8s.NewAPI(kubeconfigPath, kubeContext)
+	if err != nil {
+		return nil, err
+	}
+
+	clientset, err := kubernetes.NewForConfig(kubeAPI.Config)
+	if err != nil {
+		return nil, err
+	}
+
+	c, err := flags.NewSternConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	return &logCmdOpts{
+		clientset,
+		c,
+	}, nil
+}
+
+func newCmdLogs() *cobra.Command {
+
+	flags := &logFlags{}
+
+	cmd := &cobra.Command{
+		Use:   "logs [flags]",
+		Short: "Tail logs from kubernetes resource",
+		Long:  `Tail logs from kubernetes resource.`,
+		Example: `  # Tail logs from all containers in pods that have the prefix 'linkerd-controller' in the linkerd namespace
+  linkerd logs --pod-filter linkerd-controller.* --namespace linkerd
+
+  # Tail logs from the linkerd-proxy container in the grafana pod within the linkerd control plane
+  linkerd logs --pod-filter linkerd-grafana.* --container linkerd-proxy --namespace linkerd
+`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			opts, err := newLogOptions(flags, kubeconfigPath, kubeContext)
+
+			if err != nil {
+				return err
+			}
+
+			return runLogOutput(opts)
+		},
+	}
+
+	cmd.PersistentFlags().StringVarP(&flags.containerFilter, "container", "c", "", "Regex string to use for filtering log lines by container name")
+	cmd.PersistentFlags().StringVar(&flags.labelSelector, "label-selector", "", "kubernetes label selector to retrieve logs from resources that match")
+	cmd.PersistentFlags().StringVar(&flags.containerState, "container-state", "running", "Show logs from containers that are in a specific container state")
+	cmd.PersistentFlags().StringVarP(&flags.namespace, "namespace", "n", controlPlaneNamespace, "String to retrieve logs from pods in the given namespace")
+	cmd.PersistentFlags().StringVarP(&flags.podFilter, "pod-filter", "p", "", "Regex string to use for filtering log lines by pod name")
+	cmd.PersistentFlags().DurationVarP(&flags.sinceSeconds, "since", "s", 48*time.Hour, "Duration of how far back logs should be retrieved")
+	cmd.PersistentFlags().Int64Var(&flags.tail, "tail", -1, "Last number of log lines to show for a given container. -1 does not show any previous log lines")
+	cmd.PersistentFlags().BoolVarP(&flags.timestamps, "timestamps", "t", false, "Print timestamps for each given log line")
+
+	return cmd
+}
+
+func runLogOutput(opts *logCmdOpts) error {
+
```

Comment:
Mind removing these newlines at the beginning of each func in this file? We typically don't include them (go is long enough as is 😉).